### DISCUSSION TITLE
CORS Issue with Mosapro Integration in Vue.js Application

### DISCUSSION THREAD

**Discussion by diegodmzr** [mosparo/mosparo#179] [Problems & Issues]
Hello Mosapro Team,

I hope this email finds you well. I'm currently integrating Mosapro into my Vue.js application and have encountered a CORS (Cross-Origin Resource Sharing) issue. I would greatly appreciate your assistance in resolving this problem.

**Issue Description:**
When attempting to perform a submission verification with Mosapro from my Vue.js application, I receive the following error in the browser console:

`CORS Missing Allow Origin

Blocking of a cross-origin request (Cross-Origin Request): the "Same Origin" policy prevents accessing the remote resource at http://127.0.0.1:8080/api/v1/verification/verify. Reason: CORS header "Access-Control-Allow-Origin" is missing. Status Code: 400.
`

**Actions Taken:**

    I have verified the configuration of my server at http://127.0.0.1:8080 and ensured that the appropriate CORS headers are included in the responses.

    I configured the CORS headers in the Mosapro client initialization as follows:
    
    My code:
    `verif() {
                
                console.log(this.loginForm)
                // Create a new mosparo client
                let client = new mosparo.Client(this.host, this.publicKey, this.privateKey, {
                    'Access-Control-Allow-Origin' : '127.0.0.1:8001',
                    'Allow-Origin' : '127.0.0.1:8001',
                    'Origin' : '127.0.0.1:8001'
                });

                // Replace the next two lines with your form data retrieval logic
                let formData = this.loginForm;
                const { SumbitToken, ValidationToken } = this.getMosparoTokens();
                
                // Call the verifySubmission method
                client.verifySubmission(formData, SumbitToken, ValidationToken)
                    .then((verificationResult) => {
                    if (verificationResult.isSubmittable()) {
                        // Send the email or process the data
                        console.log('Submission is valid');
                        
                    } else {
                        // Show error message
                        console.error('Submission is not valid');
                    }
                    })
                    .catch((error) => {
                    console.error('Error verifying submission', error);
                    });
                },

        
        
        getMosparoTokens(){
            // Obtenez tous les éléments avec la classe "mosparo__checkbox"
            const mosparoInputs = document.querySelectorAll('.mosparo__checkbox input');

            // Initialisez un objet pour stocker les valeurs des tokens
            const tokens = {};

            // Parcourez chaque élément d'entrée et stockez sa valeur dans l'objet tokens
            mosparoInputs.forEach(input => {
                const tokenName = input.getAttribute('name').replace('_mosparo_', ''); // Obtenez le nom du token
                const tokenValue = input.value; // Obtenez la valeur du token
                tokens[tokenName] = tokenValue; // Stockez la valeur dans l'objet tokens
            });

            // Maintenant, vous avez les valeurs des deux tokens dans l'objet "tokens"
            console.log(tokens);

            // Vous pouvez accéder aux valeurs spécifiques comme ceci :
            const submitTokenValue = tokens['submitToken'];
            const validationTokenValue = tokens['validationToken'];

            console.log('Submit Token:', submitTokenValue);
            console.log('Validation Token:', validationTokenValue);

            return {
                "SumbitToken" : submitTokenValue,
                "ValidationToken" : validationTokenValue,
            }
        },},
        beforeMount(){  
                const Resource = {
                loadCssResource: true,
                label: 'salut',
                customMessages: {
                    fr_FR: {
                        label: 'Je suis un Humain',
                        errorSpamDetected: 'Spam!'
                    }
                },
                // Ajoutez d'autres clés et valeurs selon vos besoins
            };
            const cssLink = document.createElement('link');
            cssLink.href = 'http://127.0.0.1:8080/resources/91a55e10-50de-4dbd-af03-be593903ae49.css';
            cssLink.rel = 'stylesheet';

            document.head.appendChild(cssLink);
                    var elements = document.getElementsByClassName('mosparo__label');
            for (var i = 0; i < elements.length; i++) {
                elements[i].innerHTML = 'Nouveau texte modifié';
            }
            const mosparoScript = document.createElement('script');
            mosparoScript.src = 'http://127.0.0.1:8080/build/mosparo-frontend.js';
            mosparoScript.onload = () => {
                if (window.mosparo) {
                    // Votre code d'initialisation mosparo ici
                    console.log('mosparo script loaded');
                    // const privateKey = 'nnTsYJKOVtOIribvKDWshb07YsqBSe1ONapl0yIFQok'; // Remplacez par votre valeur réelle si applicable
                    // const formData = this.loginForm;
                    var mosparoInit = document.createElement('script');
                    var scriptText = `m = new mosparo('mosparo-box', 'http://127.0.0.1:8080', '91a55e10-50de-4dbd-af03-be593903ae49','li3s6JLDVslexJoNOQBQ7p9ys9rNxBx4bfDwQuSeVXU', ${JSON.stringify(Resource)});`;
                    mosparoInit.text = scriptText;
                    document.head.appendChild(mosparoInit);
                } else {
                    console.error("mosparo n'est pas défini dans la portée globale.");
                }
            };
            document.head.appendChild(mosparoScript);
        }
    };`
    
    
    Request for Assistance:
Despite my efforts, the problem persists. I understand that CORS is an important security feature, but I'm unsure of the best approach to resolve this issue in the context of Mosapro. Could you guide me on the correct steps to ensure a seamless integration with my Vue.js application?

Thank you in advance for your assistance. Feel free to ask for any additional information you may need to assess the situation.

Best regards,

![image](https://github.com/mosparo/mosparo/assets/157109129/d2aa86e5-2114-48bc-8cd6-f6dc558707e9)
![image](https://github.com/mosparo/mosparo/assets/157109129/ca2c763b-84d9-499f-8e6f-e36f1f376d08)
![image](https://github.com/mosparo/mosparo/assets/157109129/9064c9b6-2205-4bd0-98de-f7b7d0345ec0)
![image](https://github.com/mosparo/mosparo/assets/157109129/c2e7d221-9608-4659-b879-e10526eff351)
![image](https://github.com/mosparo/mosparo/assets/157109129/4d477304-3389-46b2-8124-cb027a9ff265)

**Comment by zepich** [comment]
**(See my second answer below regarding your issue. This message is still valid for configuring the CORS header in mosparo.)**

Hi @diegodmzr 

Thank you very much for your question.

Have you put your hostname (`127.0.0.1:8080`) as the host in your mosparo project?

![image](https://github.com/mosparo/mosparo/assets/391914/f68108c1-6acf-4742-b9bb-986d8cb7beb6)

This defines the allowed CORS origins in mosparo.

Please let me know if that does not solve the issue.

Kind regards,

zepich

**Reply by zepich** [reply]
Sorry, in your screen, I saw that your origin is `127.0.0.1:8080`, so you have to define that in the "Host" setting in mosparo

**Comment by zepich** [comment]
Hi @diegodmzr 

I'm sorry I didn't read your message carefully enough.

The error happens because you try to access the verification API. The verification API does not return a CORS header since the verification API cannot be called from the frontend.

The verification must happen in the backend code of your website, for example, in your NodeJS (or PHP, Python, etc.) backend code. It must happen in the same place where you process the data (sending the email, storing the data in the database, etc.).

This is because you cannot trust the frontend since the user can manipulate the frontend with the developer tools. Additionally, you have to use the private key of your mosparo project, which is comparable to the SMTP credentials to send an email: you do not want to show these credentials in public.

Since the CORS error is only a browser issue, you should not have any issues with CORS if you try to call the API from your backend code.

As a simple example, I've built an example with ExpressJS almost two years ago, which looks like this:

```js
var express = require('express');
var router = express.Router();

var mosparo = require('mosparo-js-api-client');
const https = require("https");

/* GET home page. */
router.get('/', function(req, res, next) {
  res.render('index', {
      title: 'mosparo demo form',
      mosparoHost: process.env.MOSPARO_HOST,
      mosparoUuid: process.env.MOSPARO_UUID,
      mosparoPublicKey: process.env.MOSPARO_PUBLIC_KEY,
  });
});

router.post('/', function (req, res, next) {
    let client = new mosparo.Client(
        process.env.MOSPARO_HOST,
        process.env.MOSPARO_PUBLIC_KEY,
        process.env.MOSPARO_PRIVATE_KEY,
        {
            httpsAgent: new https.Agent({
                rejectUnauthorized: false
            })
        }
    );

    let formData = req.body;
    try {
        let verificationPromise = client.verifySubmission(formData);
        verificationPromise.then((verificationResult) => {
            if (verificationResult.isSubmittable()) {
                res.redirect('/successful');
            } else {
                res.redirect('/error');
            }
        });
    } catch (e) {
        console.log(e);
    }
});
```

The route takes the form data, sends them to mosparo, and then redirects the user to the route `/successful` or `/error` depending on the verification result.

Kind regards,

zepich

**Comment by diegodmzr** [comment]
Subject: Persistent Issue with Mosparo Integration

Dear Zepich,

I hope this message finds you well. I am writing to inform you that unfortunately, the problem I was experiencing with the Mosparo integration persists despite my efforts to resolve it.

I have carefully followed your instructions and integrated the provided code into my application, but I continue to encounter difficulties. I have double-checked the code multiple times to ensure it aligns with your recommendations, but I have been unable to resolve the issue.

My code : 
`<template>
  <div class="login" style="margin-top: 20vh; height: 100%;">
    <el-row type="flex" justify="center">
      <el-col class="is-text-align-center">
        <el-image src="/images/logo-prosoluce-better-new-transp.png" fit="cover" alt="Logo Prosoluce" />
      </el-col>
    </el-row>
    <el-row class="is-text-align-center">
      <el-col>
        <el-form
          style="margin-top: 6vh;"
          ref="loginForm"
          :model="loginForm"
          :rules="loginRules"
          class="login-form"
          auto-complete="on"
          label-position="left"
        >
          <el-form-item prop="email">
            <el-input
              v-model="loginForm.email"
              name="email"
              type="text"
              auto-complete="on"
              placeholder="Email"
              prefix-icon="el-icon-user"
              @keyup.enter.native="verif"
            />
          </el-form-item>
          <el-form-item prop="password">
            <el-input
              v-model="loginForm.password"
              :type="pwdType"
              name="password"
              auto-complete="off"
              placeholder="Mot de passe"
              prefix-icon="el-icon-lock"
              @keyup.enter.native="verif"
            >
              <el-button
                slot="append"
                icon="el-icon-view"
                @click="showPwd"
              />
            </el-input>
          </el-form-item>

          
          <div id="mosparo-box" class="mosparo__container"></div>

          <el-form-item>
            <el-button
              :loading="loading"
              type="primary"
              @click="verif"
            >
              Connexion
            </el-button>
            <el-link href="/password/reset" style="margin-top: 100%; font-style: italic; font-size: smaller;">
              Mot de passe oublié ?
            </el-link>
          </el-form-item>
        </el-form>
      </el-col>
    </el-row>
  </div>
</template>

<script>
import { validEmail } from "@/utils/validate";
const mosparo = require('@mosparo/api-client');

export default {
  name: "logIn",

  data() {
    const validateEmail = (rule, value, callback) => {
      if (!validEmail(value)) {
        callback(new Error("L’Email n’est pas correct"));
      } else {
        callback();
      }
    };

    const validatePass = (rule, value, callback) => {
      if (value.length < 4) {
        callback(new Error("Le mot de passe doit être sup. à 4"));
      } else {
        callback();
      }
    };

    return {
      host: 'http://127.0.0.1:8080',
      publicKey: 'li3s6JLDVslexJoNOQBQ7p9ys9rNxBx4bfDwQuSeVXU',
      privateKey: 'nnTsYJKOVtOIribvKDWshb07YsqBSe1ONapl0yIFQok',
      mosparoSubmitToken: '',
      mosparoValidationToken: '',
      rawHtml: "",
      loginForm: {
        email: "",
        password: ""
      },
      loginRules: {
        email: [
          {
            required: true,
            trigger: "blur",
            validator: validateEmail
          }
        ],
        password: [
          {
            required: true,
            trigger: "blur",
            validator: validatePass
          }
        ]
      },
      loading: false,
      pwdType: "password",
      redirect: undefined
    };
  },

  watch: {
    $route: {
      handler: function (route) {
        this.redirect = route.query && route.query.redirect;
      },
      immediate: true
    }
  },

  methods: {
    showPwd() {
      if (this.pwdType === "password") {
        this.pwdType = "";
      } else {
        this.pwdType = "password";
      }
    },

    verif() {
      console.log(this.loginForm);

      // Créez un nouveau client Mosparo
      let client = new mosparo.Client(
        this.host,
        this.publicKey,
        this.privateKey,
        {
          'Access-Control-Allow-Origin': '127.0.0.1:8001',
          'Allow-Origin': '127.0.0.1:8001',
          'Origin': '127.0.0.1:8001'
        }
      );

      // Remplacez les deux lignes suivantes par votre logique d'extraction des données du formulaire
      let formData = this.loginForm;
      const { SumbitToken, ValidationToken } = this.getMosparoTokens();

      // Appelez la méthode verifySubmission
      client.verifySubmission(formData, SumbitToken, ValidationToken)
        .then((verificationResult) => {
          if (verificationResult.isSubmittable()) {
            // Envoyez l'e-mail ou traitez les données
            console.log('Submission is valid');
          } else {
            // Affichez le message d'erreur
            console.error('Submission is not valid');
          }
        })
        .catch((error) => {
          console.error('Error verifying submission', error);
        });
    },

    getMosparoTokens() {
      // Obtenez tous les éléments avec la classe "mosparo__checkbox"
      const mosparoInputs = document.querySelectorAll('.mosparo__checkbox input');

      // Initialisez un objet pour stocker les valeurs des tokens
      const tokens = {};

      // Parcourez chaque élément d'entrée et stockez sa valeur dans l'objet tokens
      mosparoInputs.forEach(input => {
        const tokenName = input.getAttribute('name').replace('_mosparo_', ''); // Obtenez le nom du token
        const tokenValue = input.value; // Obtenez la valeur du token
        tokens[tokenName] = tokenValue; // Stockez la valeur dans l'objet tokens
      });

      // Maintenant, vous avez les valeurs des deux tokens dans l'objet "tokens"
      console.log(tokens);

      // Vous pouvez accéder aux valeurs spécifiques comme ceci :
      const submitTokenValue = tokens['submitToken'];
      const validationTokenValue = tokens['validationToken'];

      console.log('Submit Token:', submitTokenValue);
      console.log('Validation Token:', validationTokenValue);

      return {
        "SumbitToken" : submitTokenValue,
        "ValidationToken" : validationTokenValue,
      };
    },
  },

  beforeMount() {
    const Resource = {
      loadCssResource: true,
      label: 'salut',
      customMessages: {
        fr_FR: {
          label: 'Je suis un Humain',
          errorSpamDetected: 'Spam!'
        }
      },
    };

    const cssLink = document.createElement('link');
    cssLink.href = 'http://127.0.0.1:8080/resources/91a55e10-50de-4dbd-af03-be593903ae49.css';
    cssLink.rel = 'stylesheet';

    document.head.appendChild(cssLink);

    var elements = document.getElementsByClassName('mosparo__label');
    for (var i = 0; i < elements.length; i++) {
      elements[i].innerHTML = 'Nouveau texte modifié';
    }

    const mosparoScript = document.createElement('script');
    mosparoScript.src = 'http://127.0.0.1:8080/build/mosparo-frontend.js';
    mosparoScript.onload = () => {
      if (window.mosparo) {
        console.log('mosparo script loaded');
        var mosparoInit = document.createElement('script');
        var scriptText = `m = new mosparo('mosparo-box', 'http://127.0.0.1:8080', '91a55e10-50de-4dbd-af03-be593903ae49','li3s6JLDVslexJoNOQBQ7p9ys9rNxBx4bfDwQuSeVXU', ${JSON.stringify(Resource)});`;
        mosparoInit.text = scriptText;
        document.head.appendChild(mosparoInit);
      } else {
        console.error("mosparo n'est pas défini dans la portée globale.");
      }
    };

    document.head.appendChild(mosparoScript);
  }
};
</script>

<style scoped>
/* ... (votre style existant) ... */
</style>
`

If you have any additional advice or if you spot any obvious errors that I might have overlooked, I would be grateful for any assistance you can provide. My goal is to address this issue so that I can fully leverage Mosparo's features within my application.

Thank you sincerely for your time and assistance. I look forward to your response.

Best regards,
![image](https://github.com/mosparo/mosparo/assets/157109129/9eb6440b-dd46-4dcc-92c6-e56598cfe4e4)


[Your Name]
[Your Company/Organization]
[Contact Information]

**Reply by diegodmzr** [reply]
![image](https://github.com/mosparo/mosparo/assets/157109129/d96ac75c-fee9-4d4c-b2c1-20617de412de)
![image](https://github.com/mosparo/mosparo/assets/157109129/4c32bca2-fda2-4d3a-8143-3559858861e2)
![image](https://github.com/mosparo/mosparo/assets/157109129/73557555-43a3-471f-a161-9d46bea084ed)
![image](https://github.com/mosparo/mosparo/assets/157109129/3b84ad88-43e9-4aac-b27b-bd41966b3323)

**Reply by zepich** [reply]
Hi @diegodmzr 

As long as you see the CORS error for the verification API of mosparo, you're calling the API from the wrong 'context' (you're calling the API from the frontend instead of the backend).

Can you show me the code where you check the email address and password? You have to verify that an account for the email address exists and that the password is correct.

Where do you execute these checks and then create the session?

Kind regards,

zepich

**Comment by diegodmzr** [comment]
Dear Mosparo Team,

I wanted to quickly inform you of the latest developments regarding the Mosparo installation. After encountering issues with the Java approach, I have decided to switch to an installation method using PHP. I am pleased to announce that this new approach has proven successful, and Mosparo is now up and running.

If you have any questions or concerns, please feel free to let me know. Thank you all for your cooperation and understanding.

Best regards,

![image](https://github.com/mosparo/mosparo/assets/157109129/34ab8dfa-67cf-4ff1-bf73-d3971d125bf5)

**Reply by zepich** [reply]
Hi @diegodmzr 

Thank you very much for your update! I'm happy that this approach is working now!

Kind regards,

zepich
